ros_packages/voice_assistant/voice_assistant/token_service.py

```diff
@@ -0,0 +1,168 @@
+        response.token_active = False
+        if self.active_token:
+            response.token_active = True
+        return response
+
+    def encrypt_token_callback(
+        self, request: EncryptToken.Request, response: EncryptToken.Response
+    ):
+        token: str = request.token
+        password: str = request.password
+
+        is_successful: bool = self.encrypt_token(token, password)
+        if is_successful:
+            self.publish_token(token)
+
+        self.is_token_stored = is_successful
+        response.is_successful = is_successful
+        return response
+
+    def decrypt_token_callback(
+        self, request: DecryptToken.Request, response: DecryptToken.Response
+    ):
+        password: str = request.password
+        is_successful: bool = True
+
+        try:
+            token: str = self.decrypt_token(password)
+            self.publish_token(token)
+        except Exception:
+            self.get_logger().warn(f"Token could not be decrypted")
+            is_successful = False
+
+        response.is_successful = is_successful
+        return response
+
+    def _check_if_previous_token_stored(self) -> bool:
+        return os.path.isfile(TOKEN_PATH)
+
+    def _generate_salt(self, size: int = 16) -> bytes:
+        return secrets.token_bytes(size)
+
+    def _generate_key(self, salt: bytes, password: str) -> bytes:
+        encoded_password: bytes = str.encode(password)
+        key = scrypt(encoded_password, salt=salt, n=2**14, r=8, p=1, dklen=32)
+        return urlsafe_b64encode(key)
+
+    def _store_secret(self, secret: bytes, filename: str) -> None:
+        if not os.path.exists(SECRETS_DIR):
+            os.makedirs(SECRETS_DIR)
+
+        with open(filename, "wb") as file:
+            file.write(secret)
+
+    def _load_secret(self, filename: str) -> bytes:
+        with open(filename, "rb") as file:
+            data = file.read()
+        return data
+
+    def encrypt_token(self, token: str, password: str) -> bool:
+        if len(password) < 8:
+            return False
+
+        try:
+            encoded_token: bytes = str.encode(token)
+            salt: bytes = self._generate_salt()
+            key: bytes = self._generate_key(salt, password)
+        except Exception as e:
+            self.get_logger().error(f"Could not generate salt/key: {e}")
+            return False
+
+        try:
+            f = Fernet(key)
+            encrypted_token: bytes = f.encrypt(encoded_token)
+        except Exception as e:
+            self.get_logger().error(f"Could not encrypt token: {e}")
+            return False
+
+        try:
+            self._store_secret(secret=salt, filename=SALT_PATH)
+            self._store_secret(secret=encrypted_token, filename=TOKEN_PATH)
+        except Exception as e:
+            self.get_logger().error(f"Could not store secrets: {e}")
+            return False
+
+        return True
+
+    def decrypt_token(self, password: str) -> str:
```

Comment:
I didn't underscore it, because it marks the "entry-point" to the encryption/decryption. In theory, one could call these methods without going through the ROS callback and could perform the complete encrypting/decrypting action. The other methods however are pure helper methods and shouldn't be used outside the class.